SAML identity provider docs. (#22414)
gravitational/teleport: docs/pages/access-controls/idps/saml.mdx

---

> title: Using Teleport as a SAML identity provider

Reviewer: This page kind of seems stuck halfway between being a guide and being a reference page.

In places, it reads like a guide, because it mentions steps in a specific sequence. It's missing some of the guidelines we use for guides though (namely the 'Step 1/N' section titles).

I would consider breaking this up into two separate pages:

- A reference page with an overview of the feature, why you might want to use it, and how it works.
- A step-by-step guide for setting it up with a particular sample app.

This would allow you to be specific in the guide and not say things like "the process varies greatly, so consult the documentation of your app"

Author: I'll do this!

description: How to configure and use Teleport as a SAML identity provider.
h1: Teleport as a SAML identity provider
---

## How to use Teleport as an identity provider

This guide will cover how to use Teleport as an identity provider so that Teleport
can be used to authenticate to external services. 

## Prerequisites

> - Teleport role with access to maintaining `saml_idp_service_provider` resources.  This is available in the default `editor` role.

Reviewer: Can we be more explicit here? What does `maintaining` mean?

Reviewer 2: I agree "maintaining" is confusing. Can we just get rid of this pre-requisite and make it a part of the guide? Show an example role that has required permissions and say that builtin "editor" role already has one.

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

> ## Configuring the SAML identity provider service

Reviewer 2: Can you follow the same pattern as other guides i.e. using step numbers.

Step 1/3. Configure SAML IdP service
Step 2/3. Establish trust relationship ...

Reviewer 2: Also, if configuration is not necessary, why do we need this step?

Author: I was more or less trying to emulate the other `sso` guides. :-) I can remove this.

> The SAML identity provider service is enabled by default on the proxy service

Reviewer: Is there a way to disable it if you don't want this functionality?

Author: There is. Talking with @alexfornuto, this is better placed in the config reference, which I can add.

for all Teleport users, so no configuration is necessary.

## Establish a trust relationship between an application and Teleport

You'll need to establish a trust relationship between an application, which will
be functionining here as a **SAML service provider**, and Teleport's SAML identity
provider. This process can vary greatly betwen applications, so you will likely
need to refer to the target application's documentation in order to figure out how to
add Teleport's identity provider metadata to the application.

### Obtain the service provider metadata and add it to Teleport

> First, you must retrieve the service provider metadata in XML format from the

Reviewer 2: Is there a well-known application that supports SAML we could use as an example?

Reviewer: Perhaps https://sptest.iamshowcase.com ??

Author: I used `samltest.id`. Let me know what you think.

application. Please refer to the application's documentation on how to do this.
Add this XML to a new file called `saml-sp.yaml`:

```yaml
kind: saml_idp_service_provider
version: v1
metadata:
  # The friendly name of the service provider. This is used to manage the
  # service provider as well as in identity provider initiated SSO.
  name: saml-service-provider
spec:
  # The entity_descriptor is the service provier XML.
  entity_descriptor: |
    <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"...
```

Add this to Teleport using `tctl`:

```code
$ tctl create saml-sp.yaml
# SAML IdP service provider 'saml-service-provider' has been created.
```

> With this, Teleport now trusts the service provider.

Reviewer 2: ```suggestion
Teleport now trusts the service provider.
```

### Configure the application for SAML SSO

The first step in configuring the application for SSO is retrieving Teleport's

> Teleport's SAML identity provider metadata. You can obtain this metadata in XML

Reviewer: ```suggestion
SAML identity provider metadata. You can obtain this metadata in XML
```

duplicate "Teleport's"

application. At present, these assertions are:

| Friendly name | Full name | Description | Format |
|---------------|-----------|-------------|--------|
| `uid` | `urn:oid:0.9.2342.19200300.100.1.1` | The user ID from Teleport | `urn:oasis:names:tc:SAML:2.0:attrname-format:uri` |
| `eduPersonAffiliation` | `urn:oid:1.3.6.1.4.1.5923.1.1.1.1` | The user's Teleport roles | `urn:oasis:names:tc:SAML:2.0:attrname-format:uri` |

Additionally, you will need to upload Teleport's identity provider metadata into
the application. Please refer to the application's documentation on how to do
this using the previously retrieved `teleport-metadata.xml`.

<Admonition type="tip">
  Many applications will allow you to upload the identity provider metadata
  directly, but some may require you to parse information out of the XML
  and add it to the application piecemeal. The application documentation will
  assist with this.
</Admonition>

Once the application is properly configured, you are ready to use Teleport as a
SAML identity provider.

## Service provider/identity provider initiated SSO

The Teleport SAML identity provider supports both service provider initiated and

> identity provider initiated SSO. Service provider initiated SSO is simple: attempt
> to log in from the configured external application. To trigger identity provider

Reviewer 2: ```suggestion
identity provider initiated SSO. Service provider initiated SSO is is triggered upon
login attempt from the configured external application. To trigger identity provider
```

initiated SSO, navigate to
`https://<proxy-address>/enterprise/saml-idp/login/<sp-name>` where
`sp-name` is the name of the service provider in Teleport. For the example above,
assuming our proxy address is `example.com`, the URL would be

> `https://example.com/enterprise/saml-idp/login/saml-service-provider`.

Reviewer 2: After this sentence, might be worth mentioning what actually happens if you go directly to this URL.

## Troubleshooting

> ### Error when logging into external application

Reviewer: I don't find this section very helpful.

Typically I would expect a troubleshooting section to list specific error scenarios and how to resolve them. As written, this basically says "if you see an error check the logs and try to fix it."

Author: Yeah, agreed. I was definitely trying to mimic the `SSO` troubleshooting sections here, but I'll get some more specific problem scenarios.
